Replace IsomorphicPerformance with a performanceNow() method

TommyBrosman opened the pull request:
IsomorphicPerformance is an internal API in client-utils. Currently it includes a number of methods supported by the [Performance](https://developer.mozilla.org/en-US/docs/Web/API/Performance) API, including `mark()` and `measure()`. The only member of the API that is currently used is `now()`. This change removes the other members and reduces the amount of work needed to restructure client-utils. Note that there is another copy of IsomorphicPerformance in common-utils. The common-utils version is deprecated, so this PR does not change it.

TommyBrosman committed:
Removed all APIs except for `now()` from performanceIsomorphic. Builds if type tests are disabled for client-utils.

TommyBrosman committed:
Fixed the build by adding the appropriate typeValidation.broken entries.

alexvy86 commented:
> Note that there is another copy of IsomorphicPerformance in common-utils. The common-utils version is deprecated, so this PR does not change it.

Yeah, common-utils used to be the home for this and got replaced with client-utils. Common-utils should be removed from the repo eventually; can't recall if there's anything that still depends on it (or its presence). Some server code might, and at this point it should probably just grab the bits it needs and move them to the server packages.

alexvy86 commented on packages/common/client-utils/src/performanceIsomorphic.ts:
```diff
@@ -9,8 +9,7 @@
  */
-export type IsomorphicPerformance = Partial<Performance> &
-	Pick<Performance, "clearMarks" | "mark" | "measure" | "now">;
+export type IsomorphicPerformance = Partial<Performance> & Pick<Performance, "now">;
```
Should we take the chance to just define `now() => number` manually now and get rid of the dependency on the `Performance` type altogether? Hadn't thought about it, but it might even help address the problem I noticed [here](https://github.com/microsoft/FluidFramework/pull/21965#issuecomment-2495109017) with our custom `performance` object not behaving like the native one when combined with `Sinon.fakeTimers`.

TommyBrosman replied:
I like that idea. The fewer environment-specific types we expose (even as an internal API), the better.

TommyBrosman replied:
Looks like there is a dependency on `getEntriesByType` in driver-base. That `Partial` includes all of the Performance fields as optional. I forgot it did this, and unless I can remove the `Partial` usage this change doesn't reduce the API for IsomorphicPerformance that much. I can manually craft a type, but it won't be compatible with the actual Performance implementation unless I either a) write isomorphic types for `PerformanceEntry` or b) keep using the environment-specific types in the signature. (a) is cleaner at a type level. Here is the build error I see:

```
@fluidframework/driver-base: error during command 'tsc --project ./tsconfig.json' (exit code 2)
@fluidframework/driver-base: src/driverUtils.ts:51:33 - error TS2339: Property 'getEntriesByType' does not exist on type 'IsomorphicPerformance'.
@fluidframework/driver-base:
@fluidframework/driver-base: 51  const resources1 = performance.getEntriesByType?.("resource") ?? [];
@fluidframework/driver-base:                                    ~~~~~~~~~~~~~~~~
@fluidframework/driver-base:
@fluidframework/driver-base:
@fluidframework/driver-base: Found 1 error in src/driverUtils.ts:51
```

`getEntriesByType` is kind of annoying because the signature exposes other types (like `PerformanceEntry`).

Side note: driverUtils.ts is doing something sketchy to avoid specifying a type here.

```ts
	// getEntriesByType is only available in browser performance object
	const resources1 = performance.getEntriesByType?.("resource") ?? [];
	// Usually the latest fetch call is to the end of resources, so we start from the end.
	for (let i = resources1.length - 1; i > 0; i--) {
		const indResTime = resources1[i] as PerformanceResourceTiming;
		const resource_name = indResTime.name.toString();
```

TommyBrosman replied:
Another thing to note is that `PerformanceResourceTiming` inherits the type that `Performance` exposes (`PerformanceEntry`). The cast would be necessary in that case as well. I'd really like to have an API where the return value of `performance.getEntriesByType?.("resource")` doesn't have the `any` type. Just writing out the types manually instead of referencing existing type packages won't solve that problem.

TommyBrosman replied:
Fixed by updating the one usage of `getEntriesByType`.

TommyBrosman committed:
- Removed the Partial usage from IsomorphicPerformance so it no longer includes the other members.
- Updated driverUtils to use `globalThis.performance` directly and avoid using IsomorphicPerformance for type information. This is a one-off case that makes use of `getEntriesByType()`.

TommyBrosman committed:
Removed the IsomorphicPerformance type and switched from `performance.now()` to `performanceNow()`.

alexvy86 reviewed (approved):
🥳

TommyBrosman committed:
Merge branch 'main' into min-iso-perf

jason-ha commented on packages/common/client-utils/src/performanceIsomorphic.ts:
```diff
@@ -4,20 +4,9 @@
 /**
- * This type contains all browser performance properties as optional, and some
- * of the intersecting properties of node and browser performance as required.
+ * Exposes the common browser performance properties used by client packages, which consists
+ * of the `now` method.
```
Comment seems off. Doesn't expose common browser performance properties.

TommyBrosman replied:
Ah nice catch. I half-updated it and missed that part.

TommyBrosman replied:
I've updated the comment to reflect the links Matt posted about the performance attribute. @jason-ha and @ChumpChief : can you please take a look? I'd like this to be the kind of API comment a new-ish person can look at and understand.

TommyBrosman replied:
Link to diff: https://github.com/microsoft/FluidFramework/pull/23668/commits/d7d506256654604c1fc882f892a80493d55ca66c#diff-391a4b1b33f4c231b40a45fc62536a66620987ce401d293d73e65cb4750ecbb2

jason-ha replied:
I might reflow the performance API portion of notes to be `@remarks` or `@privateRemarks` but works for me.

TommyBrosman replied:
Good point. I updated it.

jason-ha commented on packages/drivers/driver-base/src/driverUtils.ts:
```diff
@@ -48,7 +47,7 @@ export function getW3CData(url: string, initiatorType: string) {
 	let w3cStartTime: number | undefined; // W3C Start time = fetchStart time
 
 	// getEntriesByType is only available in browser performance object
-	const resources1 = performance.getEntriesByType?.("resource") ?? [];
+	const resources1 = globalThis.performance.getEntriesByType?.("resource") ?? [];
```
I don't know this space. `globalThis` always has `performance` property?
I tried quick documentation searches and didn't find the information I was looking for.

TommyBrosman replied:
- It is available in the Node globals, which I'm assuming globalThis points to: https://nodejs.org/api/globals.html#performance (and some documentation mentions this, though the sources aren't particularly high quality).
- As for the browser case, globalThis points to [WindowProxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis) and [Window](https://developer.mozilla.org/en-US/docs/Web/API/Window) holds a reference to the `performance` global.

ChumpChief replied:
To be precise, the property is [defined on the WindowOrWorkerGlobalScope](https://w3c.github.io/hr-time/#the-performance-attribute) (so in addition to browser vs. node, also available on workers that don't have access to `window`).

globalThis maps to the [appropriate global object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis) for the execution context it's running in.

TommyBrosman committed:
Updated a comment.

TommyBrosman committed:
Minor tweak to a comment.

Josmithr commented on packages/common/client-utils/api-report/client-utils.beta.api.md:
```diff
@@ -6,8 +6,6 @@
 
 export { EventEmitter }
 
-export { performance_2 as performance }
```
Woo! No more lies!

Josmithr replied:
Or, well, fewer lies?

Josmithr reviewed (approved):
API changes look great!
